src/main/java/com/uid2/admin/secret/SaltRotation.java

```diff
@@ -94,86 +125,142 @@ private String calculatePreviousSalt(SaltEntry salt, boolean shouldRotate, long
+
+        if (candidateIndexes.size() <= maxIndexes) {
+            return candidateIndexes;
         }
-        return salts;
+        Collections.shuffle(candidateIndexes);
+        return candidateIndexes.subList(0, Math.min(maxIndexes, candidateIndexes.size()));
     }
 
+    private List<Integer> indexesForRotation(
+            SaltEntry[] salts,
+            long minLastUpdated,
+            long maxLastUpdated,
+            List<Integer> rotatableSaltIndexes
+    ) {
+        var candidateIndexes = new ArrayList<Integer>();
+        for (int i = 0; i < salts.length; i++) {
+            var salt = salts[i];
+            var lastUpdated = salt.lastUpdated();
+            var isInTimeWindow = minLastUpdated <= lastUpdated && lastUpdated < maxLastUpdated;
+            var isRotatable = rotatableSaltIndexes.contains(i);
 
-    private void addIndexesToRotate(List<Integer> entryIndexes,
-                                    SaltEntry[] entries,
-                                    long minLastUpdated,
-                                    long maxLastUpdated,
-                                    int maxIndexes) {
-        final List<Integer> candidateIndexes = IntStream.range(0, entries.length)
-                .filter(i -> isBetween(entries[i].lastUpdated(), minLastUpdated, maxLastUpdated))
-                .boxed()
-                .collect(toList());
-        if (candidateIndexes.size() <= maxIndexes) {
-            entryIndexes.addAll(candidateIndexes);
-            return;
+            if (isInTimeWindow && isRotatable) {
+                candidateIndexes.add(i);
+            }
         }
-        Collections.shuffle(candidateIndexes);
-        candidateIndexes.stream().limit(maxIndexes).forEachOrdered(entryIndexes::add);
+        return candidateIndexes;
     }
 
-    private static boolean isBetween(long t, long minInclusive, long maxExclusive) {
-        return minInclusive <= t && t < maxExclusive;
+    private boolean isRotatable(long nextEffective, SaltEntry salt) {
+        if (this.isRefreshFromEnabled) {
+            if (salt.refreshFrom() == null) { // TODO: remove once refreshFrom is no longer optional
+                return true;
+            }
+            return salt.refreshFrom() == nextEffective;
+        }
+
+        return true;
+    }
+
+    private void logSaltAgeCounts(String logEvent, LocalDate targetDate, Instant nextEffective, SaltEntry[] salts) {
+        var formattedDate = DateTimeFormatter.ofPattern("yyyy-MM-dd").format(targetDate);
+
+        var ages = new HashMap<Long, Long>(); // salt age to count
+        for (var salt : salts) {
+            long age = (nextEffective.toEpochMilli() - salt.lastUpdated()) / DAY_IN_MS;
+            ages.put(age, ages.getOrDefault(age, 0L) + 1);
+        }
+
+        for (var entry : ages.entrySet()) {
+            LOGGER.info("{} target-date={} age={} salts={}", logEvent, formattedDate, entry.getKey(), entry.getValue());
```

Comment:
perhaps:
salts={} target-date={} age={} num_salts={}

Reply:
What's `salts` vs `num_salts` here? logEvent?

Reply:
salts would be the set of salts "rotatable-salts", "rotated-salts", etc. How were you planning on identifying the set of salts without a logfmt label, regex?

num_salts would be the number of salts